File: 128-longest-consecutive-sequence/solution.ts
function longestConsecutive(nums: number[]): number {
  const record: Record<number, boolean> = {};
  nums.forEach((num) => (record[num] = true));

  return nums.reduce((ans, num) => {
    let now = 1;
    if (record[num]) {
      delete record[num];
      let prev = num - 1;
      while (record[prev]) {
        delete record[prev];
        prev--;
        now += 1;
      }
      let next = num + 1;
      while (record[next]) {
        delete record[next];
        next++;
        now += 1;
      }
    }
    return Math.max(ans, now);
  }, 0);
}

File: 128-longest-consecutive-sequence/solution.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> st1, st2;
        for(auto x : nums) st1.insert(x);
        int ans = 0;
        for(auto x : st1) {
            if(st2.contains(x)) continue;
            int cnt = 1, y = x + 1;
            while(st1.contains(y)) {
                st2.insert(y);
                y++, cnt++;
            }
            y = x - 1;
            while(st1.contains(y)) {
                st2.insert(y);
                y--, cnt++;
            }
            ans = max(ans, cnt);
        }
        return ans;
    }
};
